plugin/service/storage/plugin.go

```diff
@@ -156,13 +149,38 @@ func (p *StoragePlugin) OnUpdateStorageBucket(ctx context.Context, req *pb.OnUpd
 		return nil, status.Errorf(codes.InvalidArgument, "error loading persisted state: %s", err)
 	}
 
+	// Verify the incoming credentials are valid and return any errors to the
+	// user if they're not. Note this doesn't validate the credentials against
+	// AWS - it only does logical validation on the fields.
+	updatedCredentials, err := cred.GetCredentialsConfig(newBucket.GetSecrets(), newStorageAttributes.CredentialAttributes, false)
+	if err != nil {
+		return nil, err
+	}
+
 	// We will be updating credentials when one of two changes occur:
 	// 1. the new bucket has secrets for static credentials, in which case we need to delete the old static credentials
 	// 2. the new bucket has a different roleARN value for dynamic credentials
 	if newBucket.GetSecrets() != nil || newStorageAttributes.RoleArn != oldStorageAttributes.RoleArn {
-		updatedCredentials, err := cred.GetCredentialsConfig(newBucket.GetSecrets(), newStorageAttributes.CredentialAttributes, false)
+		// Ensure the incoming credentials are valid for interaction with the S3
+		// Bucket before replacing them.
+		newCredState, err := cred.NewAwsCredentialPersistedState(
+			append([]cred.AwsCredentialPersistedStateOption{
+				cred.WithCredentialsConfig(updatedCredentials),
+			}, p.testCredStateOpts...)...,
+		)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "error setting up new credential persisted state: %s", err)
+		}
+		newStorageState, err := newAwsStoragePersistedState(
+			append([]awsStoragePersistedStateOption{
+				withCredentials(newCredState),
+			}, p.testStorageStateOpts...)...,
+		)
 		if err != nil {
-			return nil, err
+			return nil, status.Errorf(codes.InvalidArgument, "error loading persisted state: %s", err)
+		}
+		if err := dryRunValidation(ctx, newStorageState, newStorageAttributes, newBucket); err != nil {
```

Comment:
I think we are missing a scenario case here to validate the IAM policy attached to a new credential type is able to perform a credential rotation if the user configured a static credential type that requires us to manage it.

Kind of like a "dry run" for credential rotations.

Reply:
Actually, I see that the credential rotation validation will still occur on line 207. 

But I feel a little confused here. I'd like to talk through one of the edge cases. I'm not sure if this if conditional block guarantees that we can return to a safe state when a user updates a credential from static to static with credential rotation enabled on both updates.

What if the `dryRunValidation` succeeds, but the credential rotation fails on line 208. The `RotateCreds` function call will invoke the `RotateKeys` function in awsutil, which will return an error if the IAM policy is not configured to create or delete user access keys.  But the old credentials should still be safe to use because the error returned on line 208 will prevent the controller from updating the secrets table.

Reply:
I think I understand why I feel confused now. We never delete the old static credentials.

Reply:
I think you have a point here.

If `dryRunValidation` succeeds on line 182, we'll go ahead with the flow and call `credState.ReplaceCreds`. 

[`ReplaceCreds` does delete the old credentials if they're static credentials from AWS and we have rotated them before.](https://github.com/hashicorp/boundary-plugin-aws/blob/b8fdcee148dae5e3f82caf3ea3fac98722193a1c/internal/credential/state.go#L154-L159)

So if `RotateCreds` fails afterwards, I think we might get in a state where the old credentials got deleted and we then error, signalling Boundary to continue using those old, now-deleted credentials.

In general, in this plugin, we should validate we have a path forward before performing any destructive operations. In this case, we're deleting before we make sure we were able to secure a new credential.

I'm unsure how to fix this at this time. Ideas are welcome 😄.

Reply:
Update: So, looking at this, it seems like this problem is already present in the [current version we're using v0.1.8](https://github.com/hashicorp/boundary-plugin-aws/blob/03e5368b71fe483a9882bfa5011b7529666811b2/plugin/service/storage/plugin.go#L164-L195). 

Given our time constraints, and the fact that this problem is already present in v0.1.8, I'd say it's not in the scope of this PR to fix this issue now. Rather, it's its own issue.

Wondering what everyone thinks about this here.

Reply:
I think that is fair, we can create a patch later.